src/transport/MpiMessageEndpoint.cpp

```diff
@@ -27,4 +27,50 @@ void sendMpiHostRankMsg(const std::string& hostIn,
         endpoint.close();
     }
 }
+
+MpiMessageEndpoint::MpiMessageEndpoint(const std::string& hostIn, int portIn)
+  : sendMessageEndpoint(hostIn, portIn)
+  , recvMessageEndpoint(portIn)
+{}
+
+void MpiMessageEndpoint::sendMpiMessage(
+  const std::shared_ptr<faabric::MPIMessage>& msg)
+{
+    // TODO - is this lazy init very expensive?
+    if (sendMessageEndpoint.socket == nullptr) {
```

Comment:
In normal circumstances, I would open both `sendMessageEndpoint` and `recvMessageEndpoint` in the constructor. However, this is very problematic for testing. That's because:
1. `mpiMessageEndpoints` are `static thread local`, thus we need to create different worlds in separate threads. 
2. If both `send` and `recv` sockets are opened in the constructor, both `recv` sockets will try to bind to the same adress + port, yielding an error (even though we are only using one, in the simplest scenario).

Reply:
If the implementation relies on TLS then unfortunately the tests should exercise that TLS. As you say, this means we'd need to create worlds in different threads. 

Re. the ports and testing, the solution is probably to make the base port configurable, then change the value in tests. In this case, each world could take a port offset as a constructor argument, set to the value we currently use by default, but overridden in tests. 

Making this configurable arguable makes these classes more flexible and reusable so I'd say it's a +1 for testing!

Reply:
As discussed offline, we'll first try with binding to a specific (protocol, address, port) tuple, and will leave port management until the multi-tenancy PR.